File: tools/test_daihui_upload.py
"""E2E test: upload all 5 Daihui CSVs, validate, commit, verify generic_records."""
import json
import sys
import time
import urllib.error
import urllib.request
import uuid
from pathlib import Path

API = "http://localhost:8000"
TENANT_KEY = "r-GwQLiW-b5j1o4iOazD4fxzWlUJd6l96WI8wyFZ2nM"
CSV_DIR = Path(r"C:\Users\gslab\Desktop\岱暉資料表\處理完畢")

FILE_TO_CODE = {
    "entry_horizontal.csv": "ENTRY",
    "inspection_horizontal.csv": "INSPECTION",
    "material_horizontal.csv": "MATERIAL",
    "production_horizontal.csv": "PRODUCTION",
    "quality_horizontal.csv": "QUALITY",
}


def request(method, path, data=None, headers=None, content_type=None):
    h = {"X-API-Key": TENANT_KEY}
    if headers:
        h.update(headers)
    if content_type:
        h["Content-Type"] = content_type
    req = urllib.request.Request(f"{API}{path}", data=data, headers=h, method=method)
    try:
        resp = urllib.request.urlopen(req, timeout=30)
        return resp.status, json.loads(resp.read())
    except urllib.error.HTTPError as e:
        body = e.read()
        try:
            return e.code, json.loads(body)
        except Exception:
            return e.code, {"raw": body.decode(errors="replace")}


def upload_csv(filename, table_code):
    filepath = CSV_DIR / filename
    file_bytes = filepath.read_bytes()
    boundary = uuid.uuid4().hex.encode()
    crlf = b"\r\n"
    body = (
        b"--" + boundary + crlf
        + b'Content-Disposition: form-data; name="table_code"' + crlf + crlf
        + table_code.encode() + crlf
        + b"--" + boundary + crlf
        + b'Content-Disposition: form-data; name="allow_duplicate"' + crlf + crlf
        + b"true" + crlf
        + b"--" + boundary + crlf
        + b'Content-Disposition: form-data; name="files"; filename="' + filename.encode() + b'"' + crlf
        + b"Content-Type: text/csv" + crlf + crlf
        + file_bytes + crlf
        + b"--" + boundary + b"--" + crlf
    )
    ct = f"multipart/form-data; boundary={boundary.decode()}"
    status, resp = request("POST", "/api/v2/import/jobs", data=body, content_type=ct)
    return status, resp


def wait_for_status(job_id, target_statuses, max_wait=30):
    deadline = time.time() + max_wait
    while time.time() < deadline:
        status, resp = request("GET", f"/api/v2/import/jobs/{job_id}")
        if resp.get("status") in target_statuses:
            return resp
        time.sleep(1)
    return resp


def validate_job(job_id):
    status, resp = request("POST", f"/api/v2/import/jobs/{job_id}/validate",
                           data=b"{}", content_type="application/json")
    return status, resp


def commit_job(job_id):
    status, resp = request("POST", f"/api/v2/import/jobs/{job_id}/commit",
                           data=b"{}", content_type="application/json")
    return status, resp


PASS = 0
FAIL = 0


def ok(msg):
    global PASS
    PASS += 1
    print(f"  [PASS] {msg}")


def fail(msg):
    global FAIL
    FAIL += 1
    print(f"  [FAIL] {msg}")


print("=" * 60)
print("  Daihui CSV Upload E2E Test")
print("=" * 60)

job_ids = {}

# --- PHASE 1: Upload ---
print("\n[Phase 1] Upload CSVs")
for filename, table_code in FILE_TO_CODE.items():
    if not (CSV_DIR / filename).exists():
        fail(f"CSV not found: {filename}")
        continue
    status, resp = upload_csv(filename, table_code)
    if status in (200, 201) and resp.get("id"):
        job_id = resp["id"]
        job_ids[filename] = job_id
        ok(f"{filename} → {table_code} → job {job_id[:8]}… (status={resp.get('status')})")
    else:
        fail(f"{filename} upload failed: HTTP {status} {resp}")

if not job_ids:
    print("\n[ABORT] No jobs created.")
    sys.exit(1)

# --- PHASE 2: Wait for auto-validation (background task) ---
print("\n[Phase 2] Wait for auto-validation (background)")
ready_jobs = {}
for filename, job_id in job_ids.items():
    job_state = wait_for_status(job_id, ("READY", "FAILED", "COMPLETED"), max_wait=30)
    js = job_state.get("status")
    ec = job_state.get("error_count", "?")
    if js == "READY":
        ready_jobs[filename] = job_id
        ok(f"{filename} → READY (error_count={ec})")
    else:
        fail(f"{filename} → {js} (expected READY, error_count={ec})")

# --- PHASE 3: Commit ---
print("\n[Phase 3] Commit jobs")
committed = []
for filename, job_id in ready_jobs.items():
    status, resp = commit_job(job_id)
    if status in (200, 202):
        job_state = wait_for_status(job_id, ("COMPLETED", "FAILED"), max_wait=20)
        if job_state.get("status") == "COMPLETED":
            committed.append(filename)
            ok(f"{filename} → COMPLETED (committed_count={job_state.get('committed_count', '?')})")
        else:
            fail(f"{filename} → {job_state.get('status')} after commit, error_summary={job_state.get('error_summary')}")
    else:
        fail(f"{filename} commit HTTP {status}: {resp}")

# --- PHASE 4: DB verification via /api/forms/{code}/records ---
print("\n[Phase 4] Verify generic_records via API")
code_for_file = {v: k for k, v in FILE_TO_CODE.items()}
for table_code in FILE_TO_CODE.values():
    status, resp = request("GET", f"/api/forms/{table_code}/records")
    if status == 200:
        count = resp.get("total") or len(resp.get("records", resp if isinstance(resp, list) else []))
        if count and int(count) > 0:
            ok(f"{table_code}: {count} record(s) in generic_records")
        else:
            fail(f"{table_code}: 0 records in generic_records (response={resp})")
    else:
        fail(f"{table_code}: GET /api/forms/{table_code}/records → HTTP {status}")

print(f"\n{'='*60}")
print(f"  PASS: {PASS}  FAIL: {FAIL}")
print(f"{'='*60}")
sys.exit(0 if FAIL == 0 else 1)

File: tools/test-daihui-upload.mjs
#!/usr/bin/env node
import { readFile } from "node:fs/promises";
import { existsSync } from "node:fs";
import path from "node:path";

const BASE_URL = process.env.BASE_URL || "http://127.0.0.1:8000";
const TENANT_API_KEY =
  process.env.TENANT_API_KEY || "r-GwQLiW-b5j1o4iOazD4fxzWlUJd6l96WI8wyFZ2nM";
const CSV_DIR =
  process.env.CSV_DIR || "C:\\Users\\gslab\\Desktop\\岱暉資料表\\處理完畢";

const FILES = [
  { filenames: ["entry.csv", "entry_horizontal.csv"], tableCode: "daihui_entry" },
  { filenames: ["inspection.csv", "inspection_horizontal.csv"], tableCode: "daihui_inspection" },
  { filenames: ["material.csv", "material_horizontal.csv"], tableCode: "daihui_material" },
  { filenames: ["production.csv", "production_horizontal.csv"], tableCode: "daihui_production" },
  { filenames: ["quality.csv", "quality_horizontal.csv"], tableCode: "daihui_quality" },
];

let pass = 0;
let fail = 0;

function ok(message) {
  pass += 1;
  console.log(`  [PASS] ${message}`);
}

function bad(message) {
  fail += 1;
  console.log(`  [FAIL] ${message}`);
}

function headers(extra = {}) {
  return { "X-API-Key": TENANT_API_KEY, ...extra };
}

async function request(method, apiPath, options = {}) {
  const response = await fetch(`${BASE_URL}${apiPath}`, {
    method,
    headers: options.headers ?? headers(),
    body: options.body,
  });
  const text = await response.text();
  let data = {};
  if (text) {
    try {
      data = JSON.parse(text);
    } catch {
      data = { raw: text };
    }
  }
  return { status: response.status, data };
}

function resolveCsvFile(filenames) {
  for (const filename of filenames) {
    const fullPath = path.join(CSV_DIR, filename);
    if (existsSync(fullPath)) return { filename, fullPath };
  }
  return { filename: filenames[0], fullPath: path.join(CSV_DIR, filenames[0]) };
}

async function uploadCsv(fileSpec) {
  const { filename, fullPath } = resolveCsvFile(fileSpec.filenames);
  if (!existsSync(fullPath)) {
    throw new Error(`CSV not found: ${fullPath}`);
  }
  const bytes = await readFile(fullPath);
  const form = new FormData();
  form.append("table_code", fileSpec.tableCode);
  form.append("allow_duplicate", "true");
  form.append("files", new Blob([bytes], { type: "text/csv" }), filename);
  const result = await request("POST", "/api/v2/import/jobs", {
    headers: headers(),
    body: form,
  });
  return { ...result, filename };
}

async function waitForStatus(jobId, statuses, timeoutMs) {
  const deadline = Date.now() + timeoutMs;
  let last = {};
  while (Date.now() < deadline) {
    const result = await request("GET", `/api/v2/import/jobs/${jobId}`);
    last = result.data;
    if (statuses.has(last.status)) return last;
    await new Promise((resolve) => setTimeout(resolve, 1000));
  }
  return last;
}

async function main() {
  console.log("=".repeat(60));
  console.log("  Daihui CSV Upload E2E Test");
  console.log("=".repeat(60));
  console.log(`Backend: ${BASE_URL}`);
  console.log(`CSV_DIR: ${CSV_DIR}`);

  try {
    const health = await request("GET", "/healthz");
    if (health.status === 200) {
      ok(`GET /healthz -> HTTP ${health.status}`);
    } else {
      bad(`GET /healthz -> HTTP ${health.status}`);
      process.exit(1);
    }
  } catch (error) {
    bad(`Backend is not reachable at ${BASE_URL}: ${error.message}`);
    process.exit(1);
  }

  const readyJobs = [];
  console.log("\n[Phase 1] Upload CSVs and wait for READY");
  for (const fileSpec of FILES) {
    try {
      const upload = await uploadCsv(fileSpec);
      if (![200, 201].includes(upload.status) || !upload.data.id) {
        bad(`${fileSpec.tableCode}: upload HTTP ${upload.status} ${JSON.stringify(upload.data)}`);
        continue;
      }
      ok(`${upload.filename} -> ${fileSpec.tableCode} job ${String(upload.data.id).slice(0, 8)}`);
      const state = await waitForStatus(upload.data.id, new Set(["READY", "FAILED", "COMPLETED"]), 30000);
      if (state.status === "READY") {
        readyJobs.push({ ...fileSpec, filename: upload.filename, jobId: upload.data.id });
        ok(`${upload.filename} -> READY (error_count=${state.error_count ?? "?"})`);
      } else {
        bad(`${upload.filename} -> ${state.status ?? "UNKNOWN"} (expected READY)`);
      }
    } catch (error) {
      bad(`${fileSpec.tableCode}: ${error.message}`);
    }
  }

  console.log("\n[Phase 2] Commit jobs");
  for (const job of readyJobs) {
    const result = await request("POST", `/api/v2/import/jobs/${job.jobId}/commit`, {
      headers: headers({ "Content-Type": "application/json" }),
      body: "{}",
    });
    if (![200, 202].includes(result.status)) {
      bad(`${job.filename}: commit HTTP ${result.status} ${JSON.stringify(result.data)}`);
      continue;
    }
    const state = await waitForStatus(job.jobId, new Set(["COMPLETED", "FAILED"]), 20000);
    if (state.status === "COMPLETED") {
      ok(`${job.filename} -> COMPLETED (committed_count=${state.committed_count ?? "?"})`);
    } else {
      bad(`${job.filename} -> ${state.status ?? "UNKNOWN"} after commit`);
    }
  }

  console.log("\n[Phase 3] Verify generic_records via API");
  for (const fileSpec of FILES) {
    const result = await request("GET", `/api/forms/${fileSpec.tableCode}/records`);
    if (result.status !== 200) {
      bad(`${fileSpec.tableCode}: GET records HTTP ${result.status}`);
      continue;
    }
    const records = Array.isArray(result.data)
      ? result.data
      : result.data.records ?? result.data.items ?? [];
    const total = Number(result.data.total ?? records.length ?? 0);
    if (total > 0) {
      ok(`${fileSpec.tableCode}: ${total} persisted record(s)`);
    } else {
      bad(`${fileSpec.tableCode}: 0 persisted records`);
    }
  }

  console.log(`\n${"=".repeat(60)}`);
  console.log(`  PASS: ${pass}  FAIL: ${fail}`);
  console.log("=".repeat(60));
  process.exit(fail === 0 ? 0 : 1);
}

main().catch((error) => {
  bad(error.stack || error.message);
  process.exit(1);
});
